### PR TITLE
Fix 52: "EEXIST:file already exists" raises during creating java project

### LINKED ISSUE #52: Create java project error "EEXIST:file already exists, mkdir 'd:\vscode-java-dependency\demo2'"
Build: 41
Plat: windows/Mac/Linux

Steps:

1. F1, open the command palette
2. Select Java: Create Java Project
3. Select the location
4. Input the project name

The result  as below:

image
The created folder details as below:

image

The normal one as below:

image

### PR DESCRIPTION
**Pull request by andxu**
…project, using Promise.all will cause the racing of creating the target folder.

Fix #52

### CODE REVIEW

File: src/controllers/projectController.ts [hunk 1]
```diff
@@ -53,12 +53,11 @@ export class ProjectController {
         const projectFile: string = path.join(basePath, projectName, ".project");
         const extensionPath: string = this.context.extensionPath;
         try {
-            await Promise.all([
```
Comment (SummerSun) [inline src/controllers/projectController.ts]: Maybe ensure the directory first and put all the copy work in one Promise.all()?

File: src/controllers/projectController.ts [hunk 2]
```diff
@@ -53,12 +53,11 @@ export class ProjectController {
...
-                fse.copy(path.join(extensionPath, "templates", "App.java.sample"), path.join(basePath, projectName, "src", "app", "App.java")),
-                fse.copy(path.join(extensionPath, "templates", `Java${javaVersion}`), path.join(basePath, projectName)),
-                fse.copy(path.join(extensionPath, "templates", ".project"), projectFile),
-                fse.ensureDir(path.join(basePath, projectName, "bin")),
-            ]);
+            await fse.ensureDir(path.join(basePath, projectName, "bin"));
```
Comment (akaroml) [inline src/controllers/projectController.ts]: These two has very different meanings. The later makes it a serial operation. I don't see a problem with Promise.all. Please keep that one.

File: src/controllers/projectController.ts (hunk 1 above)
Comment (akaroml) [reply src/controllers/projectController.ts]: I agree with @SummerSun and disagree with the proposed solution. Promise.all is well known so it does not make things harder to understand. The performance downgrade is definitely noticeable when you work with larger projects. Triggering multiple copy task and let them run in parallel will make the most of modern hardware like SSD. Running them in serial manner does not make any sense.

File: templates/App.java.sample [hunk 3]
```diff
@@ -1,8 +1,7 @@
 package app;
 
 public class App {
-	public static void main(String[] args) throws Exception {
-		System.out.println("Hello Java");
-	}
-
+    public static void main(String[] args) throws Exception {
+        System.out.println("Hello Java");
```
Comment (SummerSun) [inline templates/App.java.sample]: Any change of this code snippet?

File: src/controllers/projectController.ts [hunk 4]
```diff
@@ -50,14 +50,17 @@ export class ProjectController {
     }
 
     private async scaffoldJavaProject(basePath: string, projectName: string, javaVersion: number): Promise<boolean> {
-        const projectFile: string = path.join(basePath, projectName, ".project");
-        const extensionPath: string = this.context.extensionPath;
+        const projectRoot: string = path.join(basePath, projectName);
+        const templateRoot: string = path.join(this.context.extensionPath, "templates");
+        const projectFile: string = path.join(projectRoot, ".project");
         try {
+            fse.ensureDirSync(projectRoot);
```
Comment (SummerSun) [inline src/controllers/projectController.ts]: Nice!

File: templates/App.java.sample (hunk 3 above)
Comment (andxu) [reply templates/App.java.sample]: Replacing tab with spaces

File: src/controllers/projectController.ts (hunk 4 above)
Comment (akaroml) [inline src/controllers/projectController.ts]: Can we just await fse.ensureDir?
A sync operation blocks others especially when there is potential retry logic.

**Pull request merged**